File: lib/src/database/app_database.dart
import 'dart:async';
import 'package:floor/floor.dart';
import 'package:sqflite/sqflite.dart' as sqflite;
import 'amount_converter.dart';
import 'database_expense.dart';
import 'expense_category_converter.dart';
import 'expense_dao.dart';

part 'app_database.g.dart';

@Database(version: 1, entities: [DatabaseExpense])
@TypeConverters([AmountConverter, ExpenseCategoryConverter])
abstract class AppDatabase extends FloorDatabase {
  static AppDatabase? _instance;

  static Future<AppDatabase> getInstance() async {
    _instance ??=
        await $FloorAppDatabase.databaseBuilder("app_database.db").build();

    return _instance ?? (throw StateError('Not initialized.'));
  }

  static Future<AppDatabase> getMemoryInstance() async {
    return $FloorAppDatabase.inMemoryDatabaseBuilder().build();
  }

  ExpenseDao get expenseDao;
}

File: app/src/main/java/jaanc/cents/database/AppDatabase.kt
package jaanc.cents.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters

@Database(entities = [DatabaseExpense::class], version = 1)
@TypeConverters(
    ConverterAmount::class,
    ConverterExpenseCategory::class,
    ConverterLocalDateTime::class
)
abstract class AppDatabase : RoomDatabase() {
    companion object {
        private lateinit var instance: AppDatabase

        fun getInstance(context: Context): AppDatabase {
            return synchronized(this) {
                if (!::instance.isInitialized) {
                    instance = Room.databaseBuilder(
                        context.applicationContext,
                        AppDatabase::class.java,
                        "expenses"
                    ).build()
                }

                instance
            }
        }
    }

    abstract fun expenseDao(): DaoExpense
}